### PR TITLE
Segregate EventProcessor interface

### PR DESCRIPTION
**Pull request by rranjan3**
- Create interface for event processing for proxy model
- Refactor Ethereum event processor to abide by the
  interface
- Name changes for event processing module for Ethereum

Signed-off-by: Rajeev Ranjan <rajeev2.ranjan@intel.com>

### CODE REVIEW

File: sdk/avalon_sdk/ethereum/ethereum_event_processor.py [hunk 1]
```diff
@@ -170,16 +170,10 @@ def __init__(self, config):
...
-        while True:
-            # Check for any new event logs since the last poll
-            # on this filter. Though using events, this is not
-            # fully asynchronous.
-            for event in event_filter.get_new_entries():
-                await self.queue.put(event)
-                logging.debug("Event pushed into listener Queue")
-            await asyncio.sleep(LISTENER_SLEEP_DURATION)
+    def get_events(self, event_filter, *kargs, **kwargs):
+        """
```
Comment (danintel) [inline sdk/avalon_sdk/ethereum/ethereum_event_processor.py]: Add a comment here or remove the """ """ docstring.

File: sdk/avalon_sdk/ethereum/ethereum_event_processor.py [hunk 2]
```diff
@@ -215,54 +209,21 @@ def __init__(self, config):
...
-
-            callback(normalized_event, *kargs, **kwargs)
-            self.queue.task_done()
-
-    async def sync_handler(self, *kargs, **kwargs):
-        logging.info("Started synchronous handler to handle an event")
-        event = await self.queue.get()
-        logging.debug("Event popped from listener queue")
+    def process_and_delegate(self, event, callback, *kargs, **kwargs):
+        """
```
Comment (danintel) [inline sdk/avalon_sdk/ethereum/ethereum_event_processor.py]: Add a comment here or remove the """ """ docstring.

File: sdk/avalon_sdk/ethereum/ethereum_event_processor.py [hunk 3]
```diff
@@ -275,6 +236,12 @@ def get_listeners(self, loop, event_filter):
             return [loop.create_task(
                 self.listener(event_filter)) for _ in range(1)]
 
+    def get_handlers(self, loop, callback, *kargs, **kwargs):
+        """
```
Comment (danintel) [inline sdk/avalon_sdk/ethereum/ethereum_event_processor.py]: Add a comment here or remove the """ """ docstring.

File: sdk/avalon_sdk/interfaces/event_processor.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+LISTENER_SLEEP_DURATION = 2  # second
+
+class EventProcessor(ABC):
+
+    async def listener(self, event_filter, *kargs, **kwargs):
+        """
+        This function defines a listener that runs an infinite loop
+        to listen for events from an event source. It might use pass
+        on the event_filter to the blockchain specific get_events()
+        fucntion call. It puts the raw events into a shared queue
```
Comment (danintel) [inline sdk/avalon_sdk/interfaces/event_processor.py]: change fucntion to function

File: sdk/avalon_sdk/interfaces/event_processor.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        event processor. It processes the event passed in to it and then
+        passes on the processed event to the callback function for
+        appropriate handling.
+        """
+        pass
+
+    async def sync_handler(self, *kargs, **kwargs):
+        """
+        Unlike handler(), this function handles only one event at a time.
+        It makes a call to process_and_get() with the event receied which
```
Comment (danintel) [inline sdk/avalon_sdk/interfaces/event_processor.py]: change receied to received

File: sdk/avalon_sdk/interfaces/event_processor.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+LISTENER_SLEEP_DURATION = 2  # second
+
+class EventProcessor(ABC):
+
+    async def listener(self, event_filter, *kargs, **kwargs):
+        """
+        This function defines a listener that runs an infinite loop
+        to listen for events from an event source. It might use pass
+        on the event_filter to the blockchain specific get_events()
```
Comment (danintel) [inline sdk/avalon_sdk/interfaces/event_processor.py]: I would reword this. Change:
`It might use pass on the event filter to ...`
to
`The optional event_filter paramter is passed to ...`

File: sdk/avalon_sdk/interfaces/event_processor.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        This function is meant to be implemented by blockchain specific
+        event processor. It processes the event passed in to it.
+        Returns:
+            The processed event to the synchronous event handler.
+        """
+        pass
+
+    async def start(self, event_filter, callback, *kargs, **kwargs):
+        """
+        This functions starts the event processor which boils down to
```
Comment (danintel) [inline sdk/avalon_sdk/interfaces/event_processor.py]: change functions to function

File: sdk/avalon_sdk/interfaces/event_processor.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        creating listener and handler asyncio tasks that run infinitely.
+        The listeners and handlers manage events over a shared queue which
+        is also instantiated here. This function then waits over the listeners
+        to finish using a gather() which is not possible without an external
+        factor(as the listeners are in infinite loops).
```
Comment (danintel) [inline sdk/avalon_sdk/interfaces/event_processor.py]: Add a space before `(`

File: sdk/avalon_sdk/interfaces/event_processor.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+    async def stop(self):
+        """
+        This function graciously ends the listener and handler tasks.
+        """
+        for process in self.listeners:
+            process.cancel()
+        for process in self.handlers:
+            process.cancel()
+        logging.debug("---exit---")
```
Comment (danintel) [inline sdk/avalon_sdk/interfaces/event_processor.py]: Add a newline to the end of this file

**Comment by rranjan3** [comment]
> I have several corrections that need to be made, but they are minor, so I am approving it.

Updated with rework for all your observations @danintel . Thanks

File: sdk/avalon_sdk/interfaces/event_processor.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+        This function defines a typical event handler that waits on
+        a shared queue to get events in an infinite loop. Once it gets
+        an event it passes it on to process_and_delegate() which
+        thereafter might process and delegate to the callback function.
+        """
+        logging.info("Started handler to handle events")
+        while True:
+            event = await self.queue.get()
+            logging.debug("Event popped from listener queue")
+            self.process_and_delegate(event, callback, *kargs, **kwargs)
```
Comment (manojgop) [inline sdk/avalon_sdk/interfaces/event_processor.py]: Need to define a way to exist this function. For example, process_and_delegate() return value could be checked and break from this loop.

File: sdk/avalon_sdk/interfaces/event_processor.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+            self.queue.task_done()
+
+    @abstractmethod
+    def process_and_delegate(self, event, callback, *kargs, **kwargs):
+        """
+        This function is meant to be implemented by blockchain specific
+        event processor. It processes the event passed in to it and then
+        passes on the processed event to the callback function for
+        appropriate handling.
```
Comment (manojgop) [inline sdk/avalon_sdk/interfaces/event_processor.py]: Not sure of the use of callback in process_and_delegate(). How to process the events is already implementations specific.

File: sdk/avalon_sdk/interfaces/event_processor.py (hunk 5 above)
Comment (manojgop) [reply sdk/avalon_sdk/interfaces/event_processor.py]: We could possibly combine `sync_handler()` and `handler()` into one. One way is to check the return value of `process_and_delegate()` to decide to exit from `handler()` function. That way we also don't need `process_and_get()`.

**Comment by rranjan3** [comment]
Closing this. Picked up here - https://github.com/hyperledger/avalon/pull/533.
